Implement operator installer revision history pruning controller

damemi opened the pull request:
JIRA: https://jira.coreos.com/browse/MSTR-221

~~Depends on https://github.com/openshift/api/pull/128~~
Now depends on https://github.com/openshift/library-go/pull/141

Prunes installer revision history based on a specified limit in the `OperatorSpec`, with `0` being unlimited.

damemi commented:
cc @mfojtik

deads2k commented on pkg/operator/staticpod/controller/installer/installer_controller.go:
```diff
@@ -244,6 +255,7 @@ func (c *InstallerController) createInstallerPod(currNodeState *operatorv1alpha1
 		fmt.Sprintf("--pod=%s", c.configMaps[0]),
 		fmt.Sprintf("--resource-dir=%s", "/etc/kubernetes/static-pod-resources"),
 		fmt.Sprintf("--pod-manifest-dir=%s", "/etc/kubernetes/manifests"),
+		fmt.Sprintf("--revision-history-limit=%d", operatorSpec.RevisionHistoryLimit),
```
I wouldn't expect an installer pod to know this.  I would expect a separate binary to deal with pruning

damemi replied:
Why can't the installer handle cleaning up old installations? It created them so I don't think it's unreasonable for it to prune them too

deads2k commented on pkg/operator/staticpod/installerpod/cmd.go:
```diff
@@ -77,6 +78,7 @@ func (o *InstallOptions) AddFlags(fs *pflag.FlagSet) {
 	fs.StringSliceVar(&o.ConfigMapNamePrefixes, "configmaps", o.ConfigMapNamePrefixes, "list of configmaps to be included")
 	fs.StringVar(&o.ResourceDir, "resource-dir", o.ResourceDir, "directory for all files supporting the static pod manifest")
 	fs.StringVar(&o.PodManifestDir, "pod-manifest-dir", o.PodManifestDir, "directory for the static pod manifest")
+	fs.IntVar(&o.RevisionHistoryLimit, "revision-history-limit", 0, "limit on installer pod attempts to preserve, 0=unlimited")
```
> "number of deployments to keep on disk" maybe?

I don't think the reaping pod should be making the choice. I think it should be given a revision number of `maxEligibleRevisionID` and a list of `protectedRevisionID` and it deletes every unprotected eligible revision.

damemi commented:
@deads2k updated this, could you take another look?

deads2k commented on pkg/operator/staticpod/controller/prune/prune_controller.go:
```diff
@@ -0,0 +1,223 @@
+	return fmt.Sprintf("prune-%s-%s", podName, nodeName)
+}
+
+func (c *PruneController) pruneRevisionHistory(operatorStatus *operatorv1.StaticPodOperatorStatus) error {
```
nit: move the action above the boilerplate

deads2k commented on pkg/operator/staticpod/controller/prune/prune_controller.go:
```diff
@@ -0,0 +1,223 @@
+func (c *PruneController) pruneRevisionHistory(operatorStatus *operatorv1.StaticPodOperatorStatus) error {
+	for _, nodeStatus := range operatorStatus.NodeStatuses {
+		// Read status config map to figure which deployments to keep
+		statusConfigMap, err := c.kubeClient.CoreV1().ConfigMaps(c.targetNamespace).Get(getStatusConfigMapName(nodeStatus.TargetRevision, nodeStatus.NodeName), metav1.GetOptions{})
```
Yeah, only one of these.

deads2k replied:
> Yeah, only one of these.

one per revision

deads2k commented on pkg/operator/staticpod/controller/installer/installer_controller.go:
```diff
@@ -266,6 +266,17 @@ func (c *InstallerController) newNodeStateForInstallInProgress(currNodeState *op
+	if err != nil {
+		return nil, err
+	}
+	statusConfigMap.Data[strconv.Itoa(int(currNodeState.TargetRevision))] = string(installerPod.Status.Phase)
```
> Data["nodeName.phase"]

Actually, just `Data["phase"]`

damemi replied:
@deads2k I think we should clarify the structure of the configmap we're using. Are you saying that each revision's configmap should only hold the phase for that revision? It seems like that would take a lot more work to build the list of successful/failed revisions if we can only get one data point from each revision's configmap.

What I'm trying to do is have each revision's configmap hold every other revision's statuses. The data would be copied over and updated by the InstallerController each time a new revision is made, and the PruneController can also clean up the configmap so it doesn't get massive. Then it only takes one api call to get the info (but we still need to sort it)

I don't know how we could use one data point per configmap without going back and loading in each revision's configmap data (from file?) If this isn't actually that expensive of an operation I'm fine with it, and it would let us sort the statuses in-place as we read them. I just want to make sure I understand your approach. Thoughts?

deads2k commented on pkg/operator/staticpod/controller/prune/prune_controller.go:
```diff
@@ -0,0 +1,223 @@
+				failedRevisionIDs = append(failedRevisionIDs, revisionID)
+			}
+		}
+		sort.Strings(succeededRevisionIDs)
```
this looks likely to sort improperly.  Add test to expose it.

soltysh replied:
I'm inclined to say we should move the sorting operation into `pruneEligibaleIDs` to get rid of that requirement from callers. You don't care if it is sorted here, either.

soltysh replied:
Don't we use `0001` etc numbering?

soltysh replied:
In that case David's concert about sorting numbers with alpha is critical.

deads2k commented on pkg/operator/staticpod/controller/prune/prune_controller.go:
```diff
@@ -0,0 +1,223 @@
+		fmt.Sprintf("--max-eligible-id=%s", maxEligibleRevision),
+		fmt.Sprintf("--protected-ids=%s", strings.Join(protectedRevisions, ",")),
+		fmt.Sprintf("--resource-dir=%s", "/etc/kubernetes/static-pod-resources"),
+		fmt.Sprintf("--pod-manifest-dir=%s", "/etc/kubernetes/manifests"),
```
didn't expect to see this.

damemi replied:
@deads2k can we trust that these directories are consistent enough that they can be hardcoded into the library code?

deads2k commented on pkg/operator/staticpod/controller/prune/prune_controller.go:
```diff
@@ -0,0 +1,223 @@
+		fmt.Sprintf("--protected-ids=%s", strings.Join(protectedRevisions, ",")),
+		fmt.Sprintf("--resource-dir=%s", "/etc/kubernetes/static-pod-resources"),
+		fmt.Sprintf("--pod-manifest-dir=%s", "/etc/kubernetes/manifests"),
+		fmt.Sprintf("--static-pod-name=%s", c.staticPodName),
```
also looks odd

damemi replied:
@deads2k similar question then, if this is being used by multiple operators this was just how I thought we could pass the name in

deads2k commented on pkg/operator/staticpod/controller/prune/prune_controller.go:
```diff
@@ -0,0 +1,223 @@
+		return err
+	}
+
+	err = c.pruneRevisionHistory(operatorStatus)
```
I don't see where you delete the configmaps from the API

deads2k commented on pkg/operator/staticpod/prune/cmd.go:
```diff
@@ -0,0 +1,115 @@
+		if _, protected := protectedIDs[file.Name()]; !protected &&
+			file.IsDir() &&
+			strings.HasPrefix(file.Name(), o.StaticPodName) &&
+			strings.Compare(file.Name(), o.nameForRevision(o.MaxEligibleRevisionID)) <= 0 {
```
alpha sort on numbers again.

soltysh commented on pkg/operator/staticpod/controller/prune/prune_controller.go:
```diff
@@ -0,0 +1,223 @@
+	required.Spec.Containers[0].Args = append(required.Spec.Containers[0].Args,
+		fmt.Sprintf("-v=%d", 4),
+		fmt.Sprintf("--max-eligible-id=%s", maxEligibleRevision),
+		fmt.Sprintf("--protected-ids=%s", strings.Join(protectedRevisions, ",")),
```
> This looks at risk of exceeding the CLI limit.

What limit you're talking about, the OS one is pretty high and I don't see any in cobra library.

damemi replied:
I agree with @deads2k that this could probably be handled better, with very high history limits we could be passing a lot here (at least with the current approach which runs 2 pods: one to prune successful revisions, where *all* failed revisions are "protected" followed by a pod to prune failed revisions, where *all* successful revisions are "protected")

soltysh replied:
How many protected you're imagining we'll have?

damemi replied:
The max would be the history limit set by the operator, if the pruner is running successfully. So it could be 10, if the history limit is 10, but if the history limit is 50 then we're passing a lot more. I don't know what, if any, the CLI limit is though

mfojtik commented on pkg/operator/staticpod/controller/installer/installer_controller.go:
```diff
@@ -66,6 +66,7 @@ const (
 )
 
 const revisionLabel = "revision"
+const statusConfigMapName = "revision-status"
```
nit: group consts

mfojtik commented on pkg/operator/staticpod/prune/cmd.go:
```diff
@@ -0,0 +1,111 @@
+	fs.StringVar(&o.StaticPodName, "static-pod-name", o.StaticPodName, "name of the static pod")
+}
+
+func (o *PruneOptions) Validate() error {
```
do we need to validate o.MaxEligibleRevisionID ?

damemi commented on pkg/operator/staticpod/controller/prune/prune_controller.go:
```diff
@@ -0,0 +1,213 @@
+			if err != nil {
+				return err
+			}
+			switch fmt.Sprintf("%s", phase) {
```
need to cast the `bytes[]` to a string though to switch on it right? what about `switch string(phase)`

damemi commented:
@mfojtik updated with all your feedback (and also added tests)

@deads2k could you double check that the logic for updating the configmap looks right?

deads2k commented on pkg/operator/staticpod/controller/installer/installer_controller.go:
```diff
@@ -179,6 +183,25 @@ func (c *InstallerController) manageInstallationPods(operatorSpec *operatorv1.Op
 	}
 
+	// Store success/fail status in the status configmap for revision history tracking
+	statusConfigMap, err := c.kubeClient.CoreV1().ConfigMaps(c.targetNamespace).Get(statusConfigMapName, metav1.GetOptions{})
```
this is one per revision, right?

deads2k commented on pkg/operator/staticpod/controller/installer/installer_controller.go:
```diff
@@ -179,6 +183,25 @@ func (c *InstallerController) manageInstallationPods(operatorSpec *operatorv1.Op
+		statusConfigMapObject := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{Namespace: c.targetNamespace, Name: statusConfigMapName},
+			Data: map[string]string{
+				"phase": "pending",
```
`InProgress`?

deads2k commented on pkg/operator/staticpod/controller/installer/installer_controller.go:
```diff
@@ -264,6 +291,12 @@ func (c *InstallerController) manageInstallationPods(operatorSpec *operatorv1.Op
 		break
 	}
 
+	statusConfigMap.Data["phase"] = phaseStatus
```
I expected to see a helper for updating status for a revision that interpreted the status update and decided what the phase should be .

damemi replied:
We want to look at the status update for each node and set phase to `failed` if any nodes are failed, otherwise set it as `success`, right? If that's the approach then can't I just interpret it here while we're already working on each node's status update?

deads2k commented on pkg/operator/staticpod/controller/prune/prune_controller.go:
```diff
@@ -0,0 +1,212 @@
+)
+
+type PruneController struct {
+	targetNamespace, staticPodName              string
```
project standard is a new line for each member

deads2k commented on pkg/operator/staticpod/controller/prune/prune_controller.go:
```diff
@@ -0,0 +1,212 @@
+	var succeededRevisionIDs, failedRevisionIDs []string
+
+	// Walk the resource directory looking for files that match our static pod name prefix
+	err := filepath.Walk(resourceDir, func(currentPath string, info os.FileInfo, err error) error {
```
The controller has no access to the filesystem where these defintions live.

deads2k replied:
> The controller has no access to the filesystem where these defintions live.

seems like this would come from inspecting the configmaps via the API

deads2k commented on pkg/operator/staticpod/controller/prune/prune_controller.go:
```diff
@@ -0,0 +1,212 @@
+	sort.Sort(sortorder.Natural(failedRevisionIDs))
+
+	// Get list of protected IDs
+	protectedSucceededRevisionIDs := succeededRevisionIDs[len(succeededRevisionIDs)-c.succeededRevisionLimit:]
```
What do negatives do?

deads2k commented on pkg/operator/staticpod/controller/prune/prune_controller.go:
```diff
@@ -0,0 +1,212 @@
+	// Get list of protected IDs
+	protectedSucceededRevisionIDs := succeededRevisionIDs[len(succeededRevisionIDs)-c.succeededRevisionLimit:]
+	protectedFailedRevisionIDs := failedRevisionIDs[len(failedRevisionIDs)-c.failedRevisionLimit:]
+	excludedIDs := append(protectedSucceededRevisionIDs, protectedFailedRevisionIDs...)
```
this does weird things in memory.  New slice and append both.

deads2k commented on pkg/operator/staticpod/controller/prune/prune_controller.go:
```diff
@@ -0,0 +1,212 @@
+		fmt.Sprintf("--static-pod-name=%s", c.staticPodName),
+	)
+
+	if _, err := c.kubeClient.CoreV1().Pods(c.targetNamespace).Create(required); err != nil && !apierrors.IsAlreadyExists(err) {
```
I expected you to need to pin this to the "correct" node and run three of them.

deads2k replied:
> I expected you to need to pin this to the "correct" node and run three of them.

and run one for every node

damemi replied:
added here: https://github.com/openshift/library-go/pull/82/files#diff-5c9aefda4344c1bce9b242c5c5f5d82eR114

deads2k commented on pkg/operator/staticpod/controllers.go:
```diff
@@ -47,13 +50,23 @@ func NewControllers(targetNamespaceName, staticPodName string, command, revision
+		staticPodName,
+		prunerCommand,
+		kubeClient,
+		failedRevisionLimit,
```
just hard code this in the operator for now.  Later you can pull it from the API

deads2k commented on pkg/operator/staticpod/prune/cmd.go:
```diff
@@ -0,0 +1,107 @@
+	for _, file := range files {
+		// If the file is a directory AND the revision is not protected
+		// AND it is a revision of the pod we want AND it is less than or equal to the max eligible revision
+		if _, protected := protectedIDs[file.Name()]; !protected &&
```
isn't the key for `protetedIDs` just a number?

damemi replied:
> isn't the key for protetedIDs just a number?
it was originally, but now the list of protected IDs is passed to the command with their full name (eg `kube-apiserver-pod-1`) which should match the directory for the revision. I build this into a `map[string]struct{}` above for easier lookup

deads2k commented:
Let's start breaking this into pieces.  Can you separate the code that creates the pruning command itself?

damemi commented:
@deads2k made some updates: moved everything regarding the new prune command to its own commit, and squashed down all prior feedback as well. 

I've kept the changes from the current round of review to their own commits to try and preserve the comments in a way that makes sense, but I can just squash those down as well if it's easier.

At this point I've gotten the pruner to work in manual testing (with 10+ successful revisions, seeing the right older ones deleted from `/etc/kubernetes/static-pod-resources`). However it seems that updating the `revision-status` configmap is itself kicking off a new revision, causing a constant loop. I'm not sure how to fix that, but it did make testing a lot easier. Probably not what we want though.

Still need to clean up a couple things including clearing out old resources from the namespace itself (like old revision-status configmaps) as well as testing with some failed revisions as well but this is progress

deads2k commented on pkg/operator/staticpod/controller/installer/installer_controller.go:
```diff
@@ -179,6 +184,23 @@ func (c *InstallerController) manageInstallationPods(operatorSpec *operatorv1.Op
 	}
 
+	// Store success/fail status in the status configmap for revision history tracking
+	statusConfigMap, err := c.kubeClient.CoreV1().ConfigMaps(c.targetNamespace).Get(statusConfigMapName, metav1.GetOptions{})
```
still looks like the wrong multiplicity

damemi replied:
This can probably be reworked, but I don't have the revision number until I start looking at the node statuses, which is where I update the configmap name to be unique for that revision: https://github.com/openshift/library-go/pull/82/files#diff-aea28fda19fe8a3c27c2cf6590e72d89R221

Then once outside the loop I save the configmap

damemi commented:
Added more changes to hopefully be able to use the dynamic client with unstructured to get revision limits from the operator config instead of hardcoding (https://github.com/openshift/library-go/pull/82/commits/d8fb037b2bc73b1adc6a56187e4660051c17c1fd)

and the matching changes in kube-apiserver-operator: https://github.com/openshift/cluster-kube-apiserver-operator/pull/154/commits/c4df34954e67b4775b200f997705793e31e1e1a5

deads2k commented on pkg/operator/staticpod/controller/installer/installer_controller.go:
```diff
@@ -77,11 +80,9 @@ const (
+// NewInstallerController creates a new installer controller.
 func NewInstallerController(
-	targetNamespace, staticPodName string,
+	targetNamespace, staticPodName, podResourcePrefix string,
```
project standard is separate lines

deads2k commented on pkg/operator/staticpod/controller/installer/installer_controller.go:
```diff
@@ -38,12 +38,15 @@ const (
 // installer pods to update the static pods on the master nodes.
 type InstallerController struct {
-	targetNamespace, staticPodName string
+	targetNamespace, staticPodName, podResourcePrefix string
```
project standard is separate lines

deads2k commented:
please open a new pull
